tutorials/6_comp_invars.md

```diff
@@ -0,0 +1,403 @@
+  3. (Non-interference) If the invariant holds for some client, the actions of
+     any other clients cannot break it.
+
+Recall that a neighbourhood is a fixed set of interacting clients. If our
+neighbourhoods are large enough, our proofs of correctness will generalize to
+any number of clients. The details of this can be found in the
+[last tutorial](5_client_data.md).
+
+## Instrumenting the Smart Contract
+
+The previous tutorials have gone into great detail on the instrumentation of
+[local safety properties](4_arbitrary_clients.md) and
+[adequacy checks](5_client_data.md). Therefore, we focus our attention on the
+compositional invariant. We have two challenges. First, we must find an adequate
+candidate predicate. Second, we must prove that this predicate is compositional.
+As of now, the selection of candidate predicates is manual. However, we present
+the selection as a mechanical process.
+
+Let's start by generating the model:
+
+  * `solc auction.sol --bundle=Manager --c-model --output-dir=auction`
+  * `cd auction ; mkdir build ; cd build`
+  * `CC=clang-10 CXX=clang++-10 cmake ..`
+  * `cmake --build . --target run-clang-format`
+
+For simplicity, let's also encode the property as a simple `C` method. The
+method takes in a `Manager` bundle, and checks the property for a given address:
+
+```cpp
+int property(struct Manager *c0, sol_address_t addr)
+{
+  sol_uint256_t bid = Read_Map_1(&(c0->user_auction.user_bids), addr);
+  sol_uint256_t maximum = c0->user_auction.user_maxBid;
+  return bid.v <= maximum.v;
+}
+```
+
+Let's also add a placeholder method for the compositional invariant. As `True`
+is always compositional, we will use that:
+
+```cpp
+int invariant(struct Manager *c0)
+{
+  return 1;
+}
+```
+
+Throughout the rest of this example, we produce several variations of the model.
+All variations are available online. The above instrumentation can be found at
+line 213 of the
+[first variation](https://github.com/ScottWe/smartace-examples/blob/master/tutorials/post-6/instrumented/cmodel_0.c).
+
+### Attempt One: The `True` Invariant
+
+We want to show that `forall clients x: property(Manager, x)` is an inductive
+invariant of the contract. If we were to tackle this directly, we would prove
+that the property held after constructing the bundle, and then continued to
+hold after each transaction. However, in the local setting, we need only prove
+that the compositional invariant implies the property. This is because the
+compositional invariant summarizes all possible clients.
+
+This gives us the following
+[model variation](https://github.com/ScottWe/smartace-examples/blob/master/tutorials/post-6/instrumented/cmodel_1.c):
+
+```cpp
+/* ... Contract initialization ... */
+
+while (sol_continue()) {
+  // [START] INSTRUMENTATION (LINE 250)
+  /* Reached upon initialization, and after each iteration. */
+  /* First we construct an arbitrary network. */
+  Write_Map_1(&contract_1->user_bids, Init_sol_address_t(3),
```

Comment:
Maybe use `manager_contract` and `auction_contract` instead of `contract_0` and `contract_1` and say that we change some names to improve readability?